feat: Upload files

DanielEJaimes opened the pull request.

DanielEJaimes committed:
feat: UploadFiles

DanielEJaimes committed:
fix: File type and INewFile

PedroChaparro commented:
Request a review from me on This PR when you've completed your job so that I get notified to review this.

PedroChaparro committed:
Merge branch 'dev' into feat/uploadfiles

PedroChaparro committed:
style: Format file and add missing comma

DanielEJaimes committed:
fix: Dialog show on click

Still in progress, i have this response

msg
:
"There was an error in the upload."
success
:
false

DanielEJaimes commented:
@PedroChaparro i'm not sure which type should be the fileContent

DanielEJaimes committed:
Merge branch 'dev' into feat/uploadfiles

PedroChaparro commented on src/services/files/upload-file.service.ts:
```diff
@@ -0,0 +1,56 @@
+  try {
+    const response = await axios.post(
+      `${ENVIRONMENT.PROXY_BASE_URL}/file/upload`,
+      req,
```
The upload endpoint doesn't expect a JSON. You should send the file via `multipart/form-data`, no encoding needed. There is an example using `Insomnia`: 

![image](https://github.com/hawks-atlanta/frontend-react/assets/62714297/113ba9f4-0678-49f3-afd5-b95ae675f723)

DanielEJaimes replied:
@PedroChaparro 
Then i send the file as a File type?
`type UploadfileRequest = {
  fileContent: File;
  fileName: string;
  location: string;
  token: string;
};`

PedroChaparro replied:
Yes.

DanielEJaimes replied:
I am trying to upload the files but the response says there are missing fields.
![image](https://github.com/hawks-atlanta/frontend-react/assets/131629166/da5a9ade-6ef8-4e7a-a954-a95eb3f9d0e9)

PedroChaparro reviewed (changes requested):
It seems that the endpoint is not ready for review. Keep it as a draft until you've completed your job.

DanielEJaimes committed:
refactor: file type

DanielEJaimes committed:
Merge branch 'dev' into feat/uploadfiles

DanielEJaimes committed:
feat: Upload Files

PedroChaparro commented on src/screens/dialogs/upload-file-dialog.tsx:
```diff
@@ -0,0 +1,82 @@
+    const fileInput = document.querySelector<HTMLInputElement>("#myfile");
+    if (fileInput?.files?.length) {
+      const fileList = fileInput.files;
+      for (let i = 0; i < fileList.length; i++) {
```
Since your input doesn't have the `multiple` attribute. It is not possible to upload multiple files and this `for` loop its not necessary. 

I suggest to add the `multiple` attribute to the input because it's required that the users are able to upload multiple files at once.

PedroChaparro commented on src/screens/dialogs/upload-file-dialog.tsx:
```diff
@@ -0,0 +1,82 @@
+  const { session } = useContext(AuthContext);
+
+  const uploadFile = async () => {
+    const fileInput = document.querySelector<HTMLInputElement>("#myfile");
```
I think it would be better if you use an `useRef` or an `state` to store the files instead of selecting the input with the `querySelector` method.

PedroChaparro commented on src/screens/dialogs/upload-file-dialog.tsx:
```diff
@@ -0,0 +1,82 @@
+        aria-label="Select the files"
+        className="w-full rounded-lg border p-2"
+      />
+      <button
```
Please, show a loader or something else to indicate that the files are being uploaded. Right now the modal wouldn't close until the files are completely uploaded and the button will be active while the files are being uploaded, so the users can send **multiple request with heavy files** which can affect our bandwidth. 

I think one option could be create a `loading` state and shown an spinner among the text in the button and add a conditional to the `uploadFile` function: 

```typescript
if(loading) return; 
```

![image](https://github.com/hawks-atlanta/frontend-react/assets/62714297/7337d365-143e-412c-befb-faed280f42a6)

PedroChaparro commented on src/services/files/upload-file.service.ts:
```diff
@@ -0,0 +1,62 @@
+      }
+    );
+
+    if (response.status !== 201) {
```
Not needed.

DanielEJaimes committed:
fix: multiple upload

DanielEJaimes committed:
fix: lint

DanielEJaimes committed:
fix:lint

PedroChaparro commented on src/components/Sidebar/Sidebar.tsx:
```diff
@@ -7,54 +7,59 @@ export function Sidebar() {
-            <span className="hidden md:inline" aria-label="Home">
-              Home
-            </span>
+    <>
```
It seems that the `Fragment` (`< />`) is not needed.

PedroChaparro commented on src/screens/dialogs/upload-file-dialog.tsx:
```diff
@@ -0,0 +1,96 @@
+
+  const uploadFile = async () => {
+    setIsUploading(true);
+    const fileInput = fileInputRef.current;
```
You didn't add the conditional. Even if the button shows a loader, the user is still able to click on it and start a new request which can affect our bandwidth.

DanielEJaimes committed:
fix: fragment and button disable

PedroChaparro committed:
feat(ui): Animate loader

PedroChaparro reviewed (approved):
Seems to be working. Squash and merge.

DanielEJaimes commented:
i am going to restore the branch to show Ali

PedroChaparro commented:
> i am going to restore the branch to show Ali

OK, changes are in the main branch though.
